src/getPastEventsForHub.ts

```diff
@@ -0,0 +1,192 @@
+import type {
+  DefaultManagerEvent,
+  ManagerEvent,
+  PastEventOptions,
+} from './definitions/event.type';
+import retry from 'async-retry';
+import { getServerConfig } from './ServerConfigParams';
+import type AsyncRetry from 'async-retry';
+
+const DEFAULT_MANAGER_EVENTS: DefaultManagerEvent[] = [
+  'RelayServerRegistered',
+  'RelayWorkersAdded',
+  'TransactionRelayed',
+  'TransactionRelayedButRevertedByRecipient',
+];
+
+const logToEvent = (
+  log: providers.Log,
+  relayHub = getRelayHub(),
+  provider = getProvider()
+) => {
+  /*
+   * Conversion from log to event is performed using ethers.js code, see:
+   * - https://github.com/ethers-io/ethers.js/blob/f97b92bbb1bde22fcc44100af78d7f31602863ab/packages/contracts/src.ts/index.ts#L976
+   * - https://github.com/ethers-io/ethers.js/blob/f97b92bbb1bde22fcc44100af78d7f31602863ab/packages/contracts/src.ts/index.ts#L562
+   */
+  const event = <TypedEvent>deepCopy(log);
+  event.removeListener = () => {
+    return;
+  };
+  event.getBlock = () => provider.getBlock(log.blockHash);
+  event.getTransaction = () => provider.getTransaction(log.transactionHash);
+  event.getTransactionReceipt = () =>
+    provider.getTransactionReceipt(log.transactionHash);
+
+  const parsedLog = relayHub.interface.parseLog(log);
+  event.event = parsedLog.name;
+  event.eventSignature = parsedLog.signature;
+  event.decode = (data: BytesLike, topics?: Array<string>) =>
+    relayHub.interface.decodeEventLog(parsedLog.eventFragment, data, topics);
+  event.args = parsedLog.args;
+  // decodeError is missing here
+
+  return event;
+};
+
+export async function getPastEventsForHub(
+  managerAddress: string,
+  { fromBlock, toBlock }: PastEventOptions,
+  names: ManagerEvent[] = DEFAULT_MANAGER_EVENTS
+): Promise<Array<TypedEvent>> {
+  log.debug(
+    `getPastEventsForHub: [${fromBlock || 'undefined'}, ${
+      toBlock || 'undefined'
+    }], (${names.join(',')})`
+  );
+
+  const logFilters = await getLogFilters(
+    managerAddress,
+    { fromBlock, toBlock },
+    names
+  );
+
+  const logs = await performLogRequests(logFilters);
+
+  return logs.map((log) => logToEvent(log));
+}
+
+export async function performLogRequests(
+  logFilters: providers.Filter[],
+  provider: providers.Provider = getProvider(),
+  opts?: AsyncRetry.Options
+) {
+  const getProviderLogsRequests = (logFilter: providers.Filter) =>
+    retry(
+      (_, attempt) => {
+        log.debug('provider.getLogs, filter', logFilter, 'attempt', attempt);
+
+        return provider.getLogs(logFilter);
+      },
+      {
+        retries: 3,
+        minTimeout: 300,
+        onRetry: (error) => {
+          log.error(error);
+        },
+        ...opts,
+      }
+    );
+  const requests = logFilters.map((logFilter) =>
+    getProviderLogsRequests(logFilter)
+  );
+  const logs = await Promise.allSettled(requests);
+
+  return logs
+    .filter((promiseResult) => promiseResult.status === 'fulfilled')
```

Comment:
I have a question here. The logFilters may be split in different blockRange like 1 -5, 5-10 and 10-15. Im assuming that during this ranges multiple events could exists and are of our interest, but if we only return the fulfilled, that means that if during a range that the query fails, the event that we are looking for exists, we will not returning it.

Reply:
the `fulfilled` filter is not for the events, but for the promises. When we call [Promise.allSettled](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/allSettled), it waits until all the promises completed; once a promise completed, it can be in two statuses: `fulfilled` or `rejected`. That filter allows us to select only the `fulfilled` promises. For the promises that reject, we don't do anything here, because we've already logged it before (just adding a log if the request fails more than 4 times)

Reply:
We had a meeting were we discussed it